graphql_client_codegen/src/lib.rs

```diff
@@ -43,83 +44,122 @@ impl Display for GeneralError {
-                    }
-                    "json" => {
-                        let parsed: graphql_introspection_query::introspection_response::IntrospectionResponse = serde_json::from_str(&schema_string)?;
-                        schema::Schema::from(parsed)
-                    }
-                    extension => return Err(GeneralError(format!("Unsupported extension for the GraphQL schema: {} (only .json and .graphql are supported)", extension)).into())
-                };
-
-                v.insert(schema).clone()
-            }
+            btree_map::Entry::Vacant(v) => v.insert(value).clone(),
         }
     };
 
-    // We need to qualify the query with the path to the crate it is part of
-    let (query_string, query) = {
-        let mut lock = QUERY_CACHE.lock().expect("query cache is poisoned");
-        match lock.entry(query_path) {
-            btree_map::Entry::Occupied(o) => o.get().clone(),
-            btree_map::Entry::Vacant(v) => {
-                let query_string = read_file(v.key())?;
-                let query = graphql_parser::parse_query(&query_string)
-                    .map_err(|err| GeneralError(format!("Query parser error: {}", err)))?
-                    .into_static();
-                v.insert((query_string, query)).clone()
+    Ok(cached)
+}
+
+fn query_document(query_string: &str) -> Result<QueryDocument, BoxError> {
+    let document = graphql_parser::parse_query(query_string)
+        .map_err(|err| GeneralError(format!("Query parser error: {}", err)))?
+        .into_static();
+    Ok(document)
+}
+
+fn get_set_query_from_file(
+    query_path: &std::path::Path,
+) -> Result<(String, QueryDocument), BoxError> {
+    get_set_cached(&QUERY_CACHE, query_path.to_path_buf(), {
+        let query_string = read_file(query_path)?;
+        let query_document = query_document(&query_string)?;
+        (query_string, query_document)
+    })
+}
+
+fn get_set_schema_from_file(schema_path: &std::path::Path) -> Result<Schema, BoxError> {
+    get_set_cached(&SCHEMA_CACHE, schema_path.to_path_buf(), {
+        let schema_extension = schema_path
+            .extension()
+            .and_then(std::ffi::OsStr::to_str)
+            .unwrap_or("INVALID");
+        let schema_string = read_file(schema_path)?;
+        match schema_extension {
+            "graphql" | "gql" => {
+                let s = graphql_parser::schema::parse_schema::<&str>(&schema_string).map_err(|parser_error| GeneralError(format!("Parser error: {}", parser_error)))?;
+                Schema::from(s)
+            }
+            "json" => {
+                let parsed: graphql_introspection_query::introspection_response::IntrospectionResponse = serde_json::from_str(&schema_string)?;
+                Schema::from(parsed)
             }
+            extension => return Err(GeneralError(format!("Unsupported extension for the GraphQL schema: {} (only .json and .graphql are supported)", extension)).into())
         }
```

Comment:
❔ Rust Q: Does this value get computed lazily if it's in `{}`?

Reply:
Oh ahah I meant this part that you addressed in another comment with `value_func`:

> I think this doesn’t actually deduplicate the work, because the function parameter gets evaluated before the function is called. It’s only the insertion that is skipped.